crypto/elliptic-curve-endomorphism.ts

```diff
@@ -0,0 +1,300 @@
+) {
+  let endoG = endomorphismProjective(g, endoBase, p);
+
+  let [s0, s1] = decompose(s, data);
+  let S0 = bigIntToBits(s0.abs);
+  let S1 = bigIntToBits(s1.abs);
+  if (s0.isNegative) g = projectiveNeg(g, p);
+  if (s1.isNegative) endoG = projectiveNeg(endoG, p);
+
+  let h = projectiveZero;
+
+  for (let i = data.maxBits - 1; i >= 0; i--) {
+    if (S0[i]) h = projectiveAdd(h, g, p);
+    if (S1[i]) h = projectiveAdd(h, endoG, p);
+    if (i === 0) break;
+    h = projectiveDouble(h, p);
+  }
+
+  return h;
+}
+
+/**
+ * Compute constants for curve endomorphism (cube roots of unity in base and scalar field)
+ *
+ * Throws if conditions for a cube root-based endomorphism are not met.
+ */
+function computeEndoConstants(
+  Field: FiniteField,
+  Scalar: FiniteField,
+  G: GroupAffine
+) {
+  let p = Field.modulus;
+  let q = Scalar.modulus;
+  // if there is a cube root of unity, it generates a subgroup of order 3
+  assert(p % 3n === 1n, 'Base field has a cube root of unity');
+  assert(q % 3n === 1n, 'Scalar field has a cube root of unity');
+
+  // find a cube root of unity in Fq (endo scalar)
+  // we need lambda^3 = 1 and lambda != 1, which implies the quadratic equation
+  // lambda^2 + lambda + 1 = 0
+  // solving for lambda, we get lambda = (-1 +- sqrt(-3)) / 2
+  let sqrtMinus3 = Scalar.sqrt(Scalar.negate(3n));
+  assert(sqrtMinus3 !== undefined, 'Scalar field has a square root of -3');
+  let lambda = Scalar.div(Scalar.sub(sqrtMinus3, 1n), 2n);
+  assert(lambda !== undefined, 'Scalar field has a cube root of unity');
+
+  // sanity check
+  assert(Scalar.power(lambda, 3n) === 1n, 'lambda is a cube root');
+  assert(lambda !== 1n, 'lambda is not 1');
+
+  // compute beta such that lambda * (x, y) = (beta * x, y) (endo base)
+  let lambdaG = affineScale(G, lambda, p);
+  assert(lambdaG.y === G.y, 'multiplication by lambda is a cheap endomorphism');
+
+  let beta = Field.div(lambdaG.x, G.x);
+  assert(beta !== undefined, 'Gx is invertible');
+  assert(Field.power(beta, 3n) === 1n, 'beta is a cube root');
+  assert(beta !== 1n, 'beta is not 1');
+
+  // confirm endomorphism at random point
+  // TODO would be nice to have some theory instead of this heuristic
+  let R = affineScale(G, Scalar.random(), p);
+  let lambdaR = affineScale(R, lambda, p);
+  assert(lambdaR.x === Field.mul(beta, R.x), 'confirm endomorphism');
+  assert(lambdaR.y === R.y, 'confirm endomorphism');
+
+  return { endoScalar: lambda, endoBase: beta };
+}
+
+/**
+ * compute constants for GLV decomposition and upper bounds on s0, s1
+ *
+ * see {@link decompose}
+ */
+function computeGlvData(q: bigint, lambda: bigint) {
+  let [[v00, v01], [v10, v11]] = egcdStopEarly(lambda, q);
+  let det = v00 * v11 - v10 * v01;
+
+  // upper bounds for
+  // |s0| <= 0.5 * (|v00| + |v01|)
+  // |s1| <= 0.5 * (|v10| + |v11|)
+  let maxS0 = ((abs(v00) + abs(v01)) >> 1n) + 1n;
+  let maxS1 = ((abs(v10) + abs(v11)) >> 1n) + 1n;
+  let maxBits = log2(max(maxS0, maxS1));
```

Comment:
it's essential that we are able to statically compute the maximum number of bits (`maxBits`) that scalars have after GLV decomposition, with a reliable upper bound. Only with this knowledge way can we leverage the gains of reduced bit lengths in a static circuit.